pkg/operator/baremetal_pod.go

```diff
@@ -37,6 +39,35 @@ var volumeMounts = []corev1.VolumeMount{
 	},
 }
 
+// we define this service monitor in code instead of a yaml manifest
+// because we only want it to exist when metal3 is deployed
+const metal3ServiceMonitorDefinition = `
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  namespace: openshift-machine-api
+  name: metal3-metrics
+  labels:
+    k8s-app: controller
+  annotations:
+    exclude.release.openshift.io/internal-openshift-hosted: "true"
+spec:
+  namespaceSelector:
+    matchNames:
+      - openshift-machine-api
+  selector:
+    matchLabels:
+      k8s-app: controller
```

Comment:
Is this not going to select the Machine API core components as well? Is there a specific label that is applied to only the metal3 components?

Reply:
There is not, right now. I'm exploring whether I can change that as part of this PR in some local changes I haven't published yet. I tried with an earlier version and got an error and undid the changes so I could make progress in other areas.

Reply:
It turns out that field is immutable, which means we have to delete and recreate the deployment to change it (at least as far as I can tell). It doesn't seem to hurt anything to have it this way for now, and we can do something about renaming when we move all of the metal3 logic into the CBO if it turns into a problem.

Reply:
My concern is that it is going to try and scrape all of the Machine API controllers, not just the metal3 ones. When you tried this out, did you notice in prometheus if it was trying to scrape more than just the one pod, it might have left some errors in the prometheus targets page showing that it couldn't scrape the machine-api-controllers pod?

Reply:
This new stuff will only be set up when the platform is "baremetal" though, right? So it will look at the cluster-api-provider-baremetal and metal3. Maybe there are other pods?

If we have to change the label, then I think we're going to have to defer fixing the bug to 4.7 or later, because we're trying to focus on building the new operator to manage metal3 and can't afford the time to build the logic to recreate the deployment with the new setting in the MAO.

Reply:
Which field you mean immutable exactly Doug? This selector is for the new service you are defining in this PR https://github.com/openshift/machine-api-operator/pull/671/files#diff-03a356cadfb070910ca0faa53a4ab4bfR313thttps://github.com/openshift/machine-api-operator/pull/671/files#diff-03a356cadfb070910ca0faa53a4ab4bfR313 so you can choose anything.

Reply:
It is not possible to change the `k8s-app` label value in the Deployment.Spec.Selector.MatchLabels field. Attempting to change that with a patch like https://paste.centos.org/view/b0f842e8 causes the MAO to log an error every time it tries to update the Deployment:

```
W0916 15:42:54.324744 1050505 recorder_logging.go:46] &Event{ObjectMeta:{dummy.16355b11bbdfeaac  dummy    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},
InvolvedObject:ObjectReference{Kind:Pod,Namespace:dummy,Name:dummy,UID:,APIVersion:v1,ResourceVersion:,FieldPath:,},
Reason:DeploymentUpdateFailed,
Message:Failed to update Deployment.apps/metal3 -n openshift-machine-api: Deployment.apps "metal3" is invalid: spec.selector: Invalid value: 
v1.LabelSelector{MatchLabels:map[string]string{"api":"clusterapi", "k8s-app":"metal3"}, 
MatchExpressions:[]v1.LabelSelectorRequirement(nil)} field is immutable,
Source:EventSource{Component:,Host:,}, 
FirstTimestamp:2020-09-16 15:42:54.324665004 -0400 EDT m=+105.798620314,
LastTimestamp:2020-09-16 15:42:54.324665004 -0400 EDT m=+105.798620314,
Count:1,Type:Warning,
EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,}
```

I suspect that means adding a different label is going to result in the same error, because it would mean changing the same field in the Deployment.

Reply:
https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#label-selector-updates

"In API version apps/v1, a Deployment's label selector is immutable after it gets created."

So yeah. This feels pretty wrong though, as requests will go to a bunch of things we aren't intending.

Reply:
They don't, though. The label selector is the same, but the *port* is not, and both are used to filter. Only the metal3 deployment matches both.